Update to modern OpenTelemetry practices+terminology (#3138)
MicrosoftDocs/architecture-center: docs/microservices/logging-monitoring.md

This article describes best practices for monitoring a microservices application that runs on Azure Kubernetes Service (AKS).

In any complex application, at some point something will go wrong. In a microservices application, you need to track what's happening across dozens or even hundreds of services. To make sense of what's happening, you must collect telemetry from the application. Telemetry can be divided into _logs_, _traces_ and _metrics_.

**Logs** are text-based records of events that occur while the application is running. They include things like application logs (trace statements) or web server logs. Logs are primarily useful for forensics and root cause analysis.

**Traces**, also called _Operations_ connect the steps of a single request across multiple calls within and across microservices. They can provide structured observability into the interactions of system components. Traces can begin early in the request process like within the UI of an application and propagate through network services and across a network of microservices that handle the request.

> - **Spans** are units of work within a trace. Each span is connected with a single trace and can be hested with other spans. They often correspond to individual _requests_ in a cross-service operation but can also define work in individual components within a service. Spans also track outbound calls from one service to another (sometimes these are called _dependency_ records).

Reviewer: ```suggestion
- **Spans** are units of work within a trace. Each span is connected with a single trace and can be nested with other spans. They often correspond to individual _requests_ in a cross-service operation but can also define work in individual components within a service. Spans also track outbound calls from one service to another (sometimes these are called _dependency_ records).
```

Reviewer 2: In addition, I think the sentence "Each span is connected with a single trace and can be hested with other spans." is a bit confusing and jargony.

> A single operation may generate several telemetry events, so if the application experiences a high volume of traffic, it is likely to get throttled. To mitigate this problem, you can perform sampling to reduce the telemetry traffic. The tradeoff is that your metrics will be less precise. For more information, see [Sampling in Application Insights](/azure/application-insights/app-insights-sampling). You can also reduce the data volume by pre-aggregating metrics &mdash; that is, calculating statistical values such as average and standard deviation, and sending those values instead of the raw telemetry. The following blog post describes an approach to using Application Insights at scale: [Azure Monitoring and Analytics at Scale](/archive/blogs/azurecat/azure-monitoring-and-analytics-at-scale).

Reviewer 2: `The tradeoff is that your metrics will be less precise.`
This is not true if the instrumentation supports pre-aggregation. When pre-aggregation is present, the main tradeoff is that you have fewer trace samples for edge case debugging and diagnostics.

Many of the Azure SDKs have been instrumented with OpenTelemetry or are in the process of implementing it.

An application developer can add manual instrumentation using the OpenTelemetry SDKs to:

- Add instrumentation where an underlying library does not provide it
- Enrich the trace context by adding Spans to expose application-specific units of work (e.g. an order loop creating a span for the processing of each order line)
- Enrich existing Spans with entity keys to enable easier tracing (e.g. add an "OrderID" key/value to the request that processes that order). These are surfaced by the monitoring tools as structured values for querying, filtering and aggregating (without parsing out log message strings or looking for combinations of log message sequences, as was commmon with a logging-first approach.)
- Propagate trace context by accessing Trace and Span attributes, injecting traceIds into responses and payloads and/or reading traceIds from incoming messages to create Requests and Spans.

Read more about instrumentation and the OpenTelemetry SDKs in the [OpenTelemetry documentation](https://opentelemetry.io/docs/concepts/instrumenting/).

### Application Insights

> Application Insights builds on OpenTelemetry, ingesting the rich data provided by instrumented libraries and apps and providing rich visualization and query support. The [Application Insights OpenTelemetry-based instrumentation](azure/azure-monitor/app/opentelemetry-enable.md) for langauges such as .NET, Java, Node.js, or Python handles the data collection.

Reviewer 3: > handles the data collection

Data collection is done by OpenTelemetry/its instrumentation libraries. The AzureMonitor exporters for OpenTelemetry, only does one thing - exports the telemetry to Azure backend, and nothing more.

Author: Thanks! I'll revise to explain this correctly

### OpenTelemetry

OpenTelemetry is a cross-industry effort to improve tracing by standardizing the interface between applications, libraries, telemetry and data collectors. When using a libray and framework instrumented with OpenTelemetry, most of the work of tracing traditionally system operations is handled by the underlying libraries, including common scenarios like:

- Logging basic request operations: start/exit time, duration
- Exceptions thrown
- Context propagation (e.g. sending a correlation ID across HTTP call boundaries)

> Instead, the base libraries and frameworks that handle these operations that handele these calls create rich interrlated Span and Trace data structures and propagate these across contexts. Before OpenTelemetry, these were usually just injected as special log messages or as proprietary data structures specific to the vendor building the monitoring tools. OpenTelemetry also encourages a richer instrumention data model than a traditional logging-first approach, and the logs are made more useful since log messages are linked to the Traces and Spans where they were generated. This often makes finding logs associated with a specific operation or request very easy.

Reviewer: ```suggestion
Instead, the base libraries and frameworks that handle these operations create rich interrelated Span and Trace data structures and propagate these across contexts. Before OpenTelemetry, these were usually just injected as special log messages or as proprietary data structures specific to the vendor building the monitoring tools. OpenTelemetry also encourages a richer instrumentation data model than a traditional logging-first approach, and the logs are made more useful since log messages are linked to the Traces and Spans where they were generated. This often makes finding logs associated with a specific operation or request very easy.
```

> Most of the Azure SDKs have been instrumented with OpenTelemetry or in the process. _TODO: Is there a doc or link on this status we should add here?_

Reviewer: cc @lmolkova

```suggestion
Most of the Azure SDKs have been instrumented with OpenTelemetry or are in the process. _TODO: Is there a doc or link on this status we should add here?_
```

An application developer can add manual instrumentation using the OpenTelemetry SDKs to:

> - Add instrumention where an underlying library does not provide it

Reviewer: ```suggestion
- Add instrumentation where an underlying library does not provide it
```

- Enrich the trace context by adding Spans to expose application-specific units of work (e.g. an order loop creating a span for the processing of each order line)
- Enrich existing Spans with entity keys to enable easier tracing (e.g. add an "OrderID" key/value to the request that processes that order). These are surfaced by the monitoring tools as structured values for querying, filtering and aggregating (without parsing out log message strings or looking for combinations of log message sequences, as was commmon with a logging-first approach.)
- Access Trace and Span attributes to inject traceIds into responses and payloads to streamline troubleshooting.

> Read more about instrumention and the OpenTelemetry SDKs in the [OpenTelemetry documentation](https://opentelemetry.io/docs/concepts/instrumenting/).

Reviewer: ```suggestion
Read more about instrumentation and the OpenTelemetry SDKs in the [OpenTelemetry documentation](https://opentelemetry.io/docs/concepts/instrumenting/).
```

### Application Insights

> Application Insights builds on OpenTelemetry, ingesting the rich data provided by instrumented libraries and apps and providing rich visualization and query support. The Application Insights agents for various platforms such as .NET, Java, or Node.js handles the data ingestion.

Reviewer 2: ```suggestion
Application Insights builds on OpenTelemetry, ingesting the rich data provided by instrumented libraries and apps and providing rich visualization and query support. The [Application Insights OpenTelemetry-based instrumentation](azure/azure-monitor/app/opentelemetry-enable.md) for langauges such as .NET, Java, Node.js, or Python handles the data collection.
```

> If you are using .NET Core, we recommend also using the [Application Insights for Kubernetes](https://github.com/microsoft/ApplicationInsights-Kubernetes) library. This library enriches Application Insights traces with additional information such as the container, node, pod, labels, and replica set. _TODO: Check on the status of this_

Reviewer 2: CC: @cijothomas @vishweshbankwar (for awareness)

Reviewer 3: We (Cijo, Vishwesh or our team) do not own the https://github.com/microsoft/ApplicationInsights-Kubernetes
